torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -224,7 +224,10 @@ void DistEngine::executeSendFunction(
     runEngineAndAccumulateGradients(autogradContext, dummyRoot, outputEdges);
 
     // Wait for all of the outstanding rpcs to complete.
-    autogradContext->clearAndWaitForOutstandingRpcs();
+    // TODO: we currently implicitly rely this stack frame not going away
+    // before the rpcs complete (e.g. if we return the future and call
+    // wait from the caller, tests will fail).
+    autogradContext->clearAndWaitForOutstandingRpcsAsync()->waitNoThrow();
```

Comment:
@jjlilley I'm wondering why it throws exception here? if it is expected, is it better to catch exception here and log? in this case, we do not need waitNoThrow() API?

Reply:
I'll get rid of it here, since there's actually no way that it will throw as-is.
The change is still using one instance of waitNoThrow() in request_callback_impl, though once we get rid of it (future change), I can GC the waitNoThrow() api :)